envi/interactive.py

```diff
@@ -1,35 +1,78 @@
+
     try:
-        import IPython.Shell
-        ipsh = IPython.Shell.IPShell(argv=[''], user_ns=lcls, user_global_ns=gbls)
+        from IPython import embed
+        # Using IPython.embed() allows ipython to load the user's normal 
+        # configuration file.  But the global namespace is not expected to be a 
+        # dictionary so we have to do some silly things to make this work
+        # namespace
+        global_ns = DummyNamespace()
+        global_ns.__dict__ = gbls
+
+        # because we are calling embed() instead of creating a shell that we 
+        # call the interact() function of later we have to set custom 
+        # configuration values no
+        config = {
+            'autocall': 2,
+        }
+
         print(intro)
-        ipsh.mainloop()
+        shelltype = STYPE_IPYTHON_NEW
 
     except ImportError as e:
         try:
-            from IPython.terminal.interactiveshell import TerminalInteractiveShell
-            ipsh = TerminalInteractiveShell()
-            ipsh.user_global_ns.update(gbls)
-            ipsh.user_global_ns.update(lcls)
-            ipsh.autocall = 2       # don't require parenthesis around *everything*.  be smart!
+            import IPython.Shell
+            ipsh = IPython.Shell.IPShell(argv=[''], user_ns=lcls, user_global_ns=gbls)
             print(intro)
-            ipsh.mainloop()
+            shelltype = STYPE_IPYTHON
+
         except ImportError as e:
             try:
-                from IPython.frontend.terminal.interactiveshell import TerminalInteractiveShell
+                from IPython.terminal.interactiveshell import TerminalInteractiveShell
                 ipsh = TerminalInteractiveShell()
                 ipsh.user_global_ns.update(gbls)
                 ipsh.user_global_ns.update(lcls)
                 ipsh.autocall = 2       # don't require parenthesis around *everything*.  be smart!
-
+                shelltype = STYPE_IPYTHON
                 print(intro)
-                ipsh.mainloop()
+
             except ImportError as e:
-                print(e)
-                shell = code.InteractiveConsole(gbls)
-                print(intro)
-                shell.interact()
+                try:
+                    from IPython.frontend.terminal.interactiveshell import TerminalInteractiveShell
+                    ipsh = TerminalInteractiveShell()
+                    ipsh.user_global_ns.update(gbls)
+                    ipsh.user_global_ns.update(lcls)
+                    ipsh.autocall = 2       # don't require parenthesis around *everything*.  be smart!
+                    shelltype = STYPE_IPYTHON
+
+                    print(intro)
+                except ImportError as e:
+                    print(e)
+                    shell = code.InteractiveConsole(gbls)
+                    shelltype = STYPE_IPYTHON
+                    print(intro)
+
+    if shelltype == STYPE_IPYTHON_NEW:
+        embed(using='', user_module=global_ns, user_ns=lcls, **config)
```

Comment:
So why have a big if/elif tree here, when you already have a decent control using the try/catch up above? Seems like it'd be a bit more intuitive to move the embed call to the first try block (so it's easier to discern that's what it corresponds to), and the like for ipsh.mainloop and shell.interact?

Reply:
this is performed so that any exceptions that occur don't show up as "while handling this other exception, this occurred...."